R/board_gcs.R

```diff
@@ -0,0 +1,262 @@
+#'   to [googleCloudStorageR::gcs_get_global_bucket()].
+#' @param prefix Prefix within this bucket that this board will occupy.
+#'   You can use this to maintain multiple independent pin boards within
+#'   a single GCS bucket. Will typically end with `/` to take advantage of
+#'   Google Cloud Storage's directory-like handling.
+#' @export
+#' @examples
+#' \dontrun{
+#' board <- board_gcs()
+#' board %>% pin_write(mtcars)
+#' board %>% pin_read("mtcars")
+#'
+#' # A prefix allows you to have multiple independent boards in the same pin.
+#' board_sales <- board_gcs("company-pins", prefix = "sales/")
+#' board_marketing <- board_gcs("company-pins", prefix = "marketing/")
+#' # You can make the hierarchy arbitrarily deep.
+#'
+#' # Pass arguments like `predefinedAcl` through the dots of `pin_write`:
+#' board %>% pin_write(mtcars, predefinedAcl = "publicRead")
+#' }
+board_gcs <- function(bucket = googleCloudStorageR::gcs_get_global_bucket(),
+                      prefix = NULL,
+                      versioned = TRUE,
+                      cache = NULL) {
+
+  check_installed("googleCloudStorageR")
+
+  # Check that have access to the bucket
+  invisible(googleCloudStorageR::gcs_get_bucket(bucket))
+
+  cache <- cache %||% board_cache_path(paste0("gcs-", bucket))
+  new_board_v1("pins_board_gcs",
+               name = "gcs",
+               bucket = bucket,
+               prefix = prefix,
+               cache = cache,
+               versioned = versioned
+  )
+}
+
+board_gcs_test <- function(...) {
+
+  skip_if_missing_envvars(
+    tests = "board_gcs()",
+    envvars = c("PINS_GCS_PASSWORD")
+  )
+
+  rlang::check_installed("sodium")
+  path_to_encrypted_json <- fs::path_package("pins", "secret", "pins-gcs-testing.json")
+  raw <- readBin(path_to_encrypted_json, "raw", file.size(path_to_encrypted_json))
+  pw <- Sys.getenv("PINS_GCS_PASSWORD", "")
+  json <- sodium::data_decrypt(
+    bin = raw,
+    key = sodium::sha256(charToRaw(pw)),
+    nonce = secret_nonce()
+  )
+  googleCloudStorageR::gcs_auth(json_file = rawToChar(json))
+
+  board_gcs("pins-dev",
+            cache = tempfile(),
+            ...
+  )
+}
+
+#' @export
+pin_list.pins_board_gcs <- function(board, ...) {
+  NA
+}
+
+#' @export
+pin_exists.pins_board_gcs <- function(board, name, ...) {
+  withr::local_options(list(googleAuthR.verbose = 4))
+  gcs_file_exists(board, name)
+}
+
+#' @export
+pin_delete.pins_board_gcs <- function(board, names, ...) {
+  for (name in names) {
+    check_pin_exists(board, name)
+    gcs_delete_dir(board, name)
+  }
+  invisible(board)
+}
+
+#' @export
+pin_versions.pins_board_gcs <- function(board, name, ...) {
+  check_pin_exists(board, name)
+  resp <- googleCloudStorageR::gcs_list_objects(
+    bucket = board$bucket,
+    prefix = paste0(board$prefix, name)
+  )
+  paths <- fs::path_split(unique(fs::path_dir(resp$name)))
+  version_from_path(map_chr(paths, ~ .x[[length(.x)]]))
```

Comment:
This is what I did to deal with the flat namespace situation; suggestions welcome for a nicer option!

Reply:
There is the concept of object versioning in GCS, does this match what pin_versions does? https://cloud.google.com/storage/docs/object-versioning

Reply:
No, one of the ideas of pins is to provide versioning, even if the data practitioner is not allowed to turn on GCS versioning because of, say, the bucket retention policy.